### PR TITLE
Create new `history` DB for version history

### PR DESCRIPTION
**Pull request by pdcp1**
This close https://github.com/responsible-ai-collaborative/aiid/issues/2041

## Features added

- [ ] new field `epoch_date_modified` on `submissions`
- [ ] All `epoch_date_modified` fields on `reports` and `submissions` now includes hh:mm:ss
- [ ] new `logReportHistory` function that creates a new entry into the `reports` collection in the `history` DB
- [ ] new `logIncidentHistory` function that creates a new entry into the `incidents` collection in the `history` DB

## Deployment environment
https://app.netlify.com/sites/pablocosta-staging/deploys

## Testing environment
https://deploy-preview-21--pablocosta-staging.netlify.app/
Admin user: pablo@botsfactory.io
pass: testUser123

## Places where we log a history change

Reports and Incidents |  
-- | --
**URL** | **Action**
/cite/10#26 | Show more on a report and Flag report
/cite/edit/?report_number=10 | Edit report
/apps/discover | Flag report
  |  
/cite/9 | Flag incident as nor related
/apps/incidents | Edit incident
/incidents/edit?incident_id=10 | Edit incident
/cite/10 | Clone incident
/incidents/new | New incident
N/A | Promote submission to incident/report

### CODE REVIEW

**Comment by cesarvarela** [comment]
Eventually, we should ditch `epoch_*` fields and use proper Mongodb dates, which are just javascript dates 🤔

**Comment by cesarvarela** [comment]
Where can I test this?

File: site/gatsby-site/cypress/e2e/unit/functions/logIncidentHistory.cy.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  editor_dissimilar_incidents: [],
+  embedding: {
+    vector: [-0.083877206262615, 0.06872937753796578],
+    from_reports: [3148, 3147],
+  },
+  tsne: {
+    x: -0.28079595740348834,
+    y: -0.3189867109991937,
+  },
+  editor: 'Daniel Atherton',
```
Comment (cesarvarela) [inline site/gatsby-site/cypress/e2e/unit/functions/logIncidentHistory.cy.js:38]: What's the difference between `editor` and `user`?

File: site/gatsby-site/cypress/e2e/unit/functions/logIncidentHistory.cy.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
+const logIncidentHistory = require('../../../../../realm/functions/logIncidentHistory');
+
+const incidentInput = {
+  incident_id: 545,
+  title: 'Chatbot Tessa gives unauthorized diet advice to users seeking help for eating disorders',
+  description:
+    'The National Eating Disorders Association (NEDA) has shut down its chatbot named Tessa after it gave weight-loss advice to users seeking help for eating disorders. The incident has raised concerns about the risks of using chatbots and AI assistants in healthcare settings, particularly in addressing sensitive issues like eating disorders. NEDA is investigating the matter, emphasizing the need for caution and accuracy when utilizing technology to provide mental health support.',
+  reports: [3103, 3104],
+  editors: ['Daniel Atherton'],
```
Comment (cesarvarela) [inline site/gatsby-site/cypress/e2e/unit/functions/logIncidentHistory.cy.js:9]: It doesn't make much difference here, but this is a user id now.

File: site/gatsby-site/cypress/e2e/unit/functions/logReportHistory.cy.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  submitters: ['Test User'],
+  text: '[](/moneywatch)\n\nBy Aimee Picchi\n\nSeptember 24, 2015 / 11: 41 AM / MoneyWatch\n\nFor Starbucks([SBUX](http://markets.cbsnews.com/SBUX/quote/)) barista Kylei Weisse',
+  plain_text:
+    '\n\nBy Aimee Picchi\n\nSeptember 24, 2015 / 11: 41 AM / MoneyWatch\n\nFor Starbucks(SBUX) barista Kylei Weisse',
+  url: 'https://www.cbsnews.com/news/is-starbucks-shortchanging-its-baristas',
+  source_domain: 'cbsnews.com',
+  language: 'en',
+  tags: [],
+  user: { link: '63320ce63ec803072c9f529c' },
+  editor: 'Test User',
```
Comment (cesarvarela) [inline site/gatsby-site/cypress/e2e/unit/functions/logReportHistory.cy.js:29]: This is a user id now.

**Comment by pdcp1** [comment]
> Where can I test this?

Oh I forgot to add the URL to the PR description: https://deploy-preview-21--pablocosta-staging.netlify.app/

File: site/gatsby-site/src/pages/incidents/new.js [hunk 4]
```diff
@@ -85,6 +96,20 @@ function NewIncidentPage() {
 
       await insertIncident({ variables: { incident: newIncident } });
 
+      // Set the user as the last modifier
+      newIncident.modifiedBy =
+        user && user.customData.first_name && user.customData.last_name ? user.id : '';
```
Comment (cesarvarela) [inline site/gatsby-site/src/pages/incidents/new.js:101]: Should we make it a requirement to have first and last names set to be able to edit stuff instead of doing this? As it is, we'll lose track of who made the change.

File: site/realm/data_sources/mongodb-atlas/history/incidents/schema.json [hunk 5]
```diff
@@ -0,0 +1,126 @@
+{
```
Comment (cesarvarela) [inline site/realm/data_sources/mongodb-atlas/history/incidents/schema.json:1]: I would like to reuse the same schema since there is a significant chance of them going out of sync.

Maybe we could write a test that compares both and checks that they are the same except for the particular fields here 🤔

File: site/realm/functions/logReportHistory.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+ * This function creates a entry into the Report History collection
+ * 
+ * @param {object} input A JSON object that contains the Report data
+ * 
+*/
+exports = async (input) => {
+
+  const reportsHistory = context.services.get('mongodb-atlas').db('history').collection("reports");
+
+  const report = input;
```
Comment (cesarvarela) [inline site/realm/functions/logReportHistory.js:12]: The payload defined below only asks for `repor_number`, but the function here assumes that the entire report is being sent 🤔

File: site/realm/functions/logIncidentHistory.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+  const incidentsHistory = context.services.get('mongodb-atlas').db('history').collection("incidents");
+
+  const {
+    AllegedDeployerOfAISystem,
+    AllegedDeveloperOfAISystem,
+    AllegedHarmedOrNearlyHarmedParties,
+    ...incident } = input;
+
+  incident['Alleged deployer of AI system'] = AllegedDeployerOfAISystem;
```
Comment (cesarvarela) [inline site/realm/functions/logIncidentHistory.js:18]: Why did you need to remove spaces here?

File: site/realm/functions/logIncidentHistory.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  incident['Alleged developer of AI system'] = AllegedDeveloperOfAISystem;
+  incident['Alleged harmed or nearly harmed parties'] = AllegedHarmedOrNearlyHarmedParties;
+
+  await incidentsHistory.insertOne(incident);
```
Comment (cesarvarela) [inline site/realm/functions/logIncidentHistory.js:22]: The payload only expects a `incident_id` but this assumes the entire incident is being sent 🤔

**Review by cesarvarela — COMMENTED** [review]
I will have some more questions tomorrow!

File: site/gatsby-site/cypress/e2e/unit/functions/logIncidentHistory.cy.js (hunk 2 above)
Comment (pdcp1) [reply site/gatsby-site/cypress/e2e/unit/functions/logIncidentHistory.cy.js:9]: Fixed

File: site/gatsby-site/cypress/e2e/unit/functions/logIncidentHistory.cy.js (hunk 1 above)
Comment (pdcp1) [reply site/gatsby-site/cypress/e2e/unit/functions/logIncidentHistory.cy.js:38]: Actually, the correct field here is `modifiedBy`. Fixed

File: site/gatsby-site/cypress/e2e/unit/functions/logReportHistory.cy.js (hunk 3 above)
Comment (pdcp1) [reply site/gatsby-site/cypress/e2e/unit/functions/logReportHistory.cy.js:29]: Actually, the correct field here is modifiedBy. Fixed

File: site/gatsby-site/src/pages/incidents/new.js (hunk 4 above)
Comment (pdcp1) [reply site/gatsby-site/src/pages/incidents/new.js:101]: This condition is to know if the user is real, not an anonymous user from Realm. Because the `user` variable is always defined on our app.
The first and last names are not needed at this time, we only need the user Id

Comment (pdcp1) [reply site/gatsby-site/src/pages/incidents/new.js]: Fixed

File: site/realm/functions/logIncidentHistory.js (hunk 7 above)
Comment (pdcp1) [reply site/realm/functions/logIncidentHistory.js:18]: Actually, we need to add the spaces in order to insert it into the DB

File: site/realm/functions/logIncidentHistory.js (hunk 8 above)
Comment (pdcp1) [reply site/realm/functions/logIncidentHistory.js:22]: The payload is the returned value, the Input type is: `"input_type": "History_incidentInsertInput",`

File: site/realm/functions/logReportHistory.js (hunk 6 above)
Comment (pdcp1) [reply site/realm/functions/logReportHistory.js:12]: The payload is the returned value, the Input type is: `"input_type": "History_reportInsertInput",`

File: site/realm/data_sources/mongodb-atlas/history/incidents/schema.json (hunk 5 above)
Comment (pdcp1) [reply site/realm/data_sources/mongodb-atlas/history/incidents/schema.json:1]: Added!

File: site/gatsby-site/src/components/discover/Actions.js [hunk 9]
```diff
@@ -9,27 +9,59 @@ import {
...
+
+    const report = transformReportData(data.report, user);
+
+    const logReport = {
+      ...report,
+      ...updated,
+    };
+
+    // Set the user as the last modifier
+    logReport.modifiedBy = user && user.providerType != 'anon-user' ? user.id : '';
```
Comment (cesarvarela) [inline site/gatsby-site/src/components/discover/Actions.js:60]: This because `anon`s can flag and it counts as a change?

File: site/gatsby-site/src/components/incidents/IncidentEditModal.js [hunk 10]
```diff
@@ -84,6 +92,13 @@ export default function IncidentEditModal({ show, onClose, incidentId }) {
         createEntityMutation
       );
 
+      updated.epoch_date_modified = getUnixTime(new Date());
+
+      // Add the current user to the list of editors
+      if (user && user.providerType != 'anon-user' && !updated.editors.link.includes(user.id)) {
```
Comment (cesarvarela) [inline site/gatsby-site/src/components/incidents/IncidentEditModal.js:98]: Could this logic (and the `transformIncidentData` call)  be done in one function? Maybe create a custom hook `useLogHistory` that returns a function and then we have to only add one line everywhere we want to log a change:

```
const logStuff = useLogHistory();

... 

await logStuff({ ...everything needed here...)
```

File: site/gatsby-site/src/utils/cite.js [hunk 11]
```diff
@@ -124,3 +124,71 @@ export const sortIncidentsByDatePublished = (incidentReports) => {
...
+  if (embedding) {
+    // eslint-disable-next-line no-unused-vars
+    const { __typename: embeddingType, ...currentEmbedding } = embedding;
+
+    result.embedding = currentEmbedding;
+  }
+
+  if (tsne) {
+    // eslint-disable-next-line no-unused-vars
+    const { __typename: tsneType, ...currentTsne } = tsne;
```
Comment (cesarvarela) [inline site/gatsby-site/src/utils/cite.js:181]: I think if you set it to `undefined` you don't need to disable the linter

File: site/realm/functions/logIncidentHistory.js (hunk 7 above)
Comment (cesarvarela) [reply site/realm/functions/logIncidentHistory.js:18]: Oh I remember, the auto-generated graphql endpoint removes spaces and uses CamelCase

File: site/gatsby-site/src/components/discover/Actions.js (hunk 9 above)
Comment (pdcp1) [reply site/gatsby-site/src/components/discover/Actions.js:60]: Yes, you can flag without login. In that case we log the change with the empty `modifiedBy` field

File: site/gatsby-site/src/utils/cite.js (hunk 11 above)
Comment (pdcp1) [reply site/gatsby-site/src/utils/cite.js]: Fixed

**Comment by cesarvarela** [comment]
@pdcp1 please fix conflicts!

File: site/gatsby-site/src/components/incidents/IncidentEditModal.js (hunk 10 above)
Comment (pdcp1) [reply site/gatsby-site/src/components/incidents/IncidentEditModal.js:98]: Issue https://github.com/responsible-ai-collaborative/aiid/issues/2138 added
